Initial server side implementation of the Job File Sync Service

tgianos opened the pull request:
An initial implementation of the protocols and services needed to sync files from an agent job directory to a Genie server

Things are still needed for long term completeness of this feature of Genie 4:
- A way to block access to log resources until a server has caught up to expected Agent high water mark
- A way to uniquely identify a grpc connection between an agent and a server in order to de-register the appropriate observer from internal state in case the same agent reattached to the same server before a cleanup could happen
- A way to either push logs from the server to long term storage (e.g. S3) while the logs are "live" or all in one push after upload to the server is complete
- The agent side implementation of this sync protocol
- A mechanism to indicate which node a job logs are currently located on (database or other)

Currently implemented is the low level service which can write bytes to random locations in files for a given job on local disk using NIO. Also the retrieval of logs via V3 API has been modified to leverage this service in order to centralize logic for log interaction.

sumitnetflix commented on genie-web/src/main/java/com/netflix/genie/web/services/impl/DiskLogServiceImpl.java:
```diff
@@ -0,0 +1,336 @@
+            }
+        } else if (Files.isDirectory(logFile)) {
+            // TODO: Perhaps this should be different exception
+            throw new IllegalArgumentException(relativePath + " is a directory not a file. Unable to update");
```
Should be same exception as the one we decide to throw at line 256

mprimi replied:
This is a user input error.
Line 256 is the unlikely event that someone is creating files inside a private folder that genie is managing.
I disagree they need to be treated the same (even if the underlying error -- file where i'm attempting to create a directory -- is the same).

mprimi replied:
Whoops, missed placed where the other was invoked, it can be more than user error.

Still think on a very abstract level that it can be appropriate for the same underlying error/state to throw different kinds of error depending on context (and example here is user input/configuration vs. runtime unexpected state).

tgianos replied:
The argument could be made that since this service exist only to write files from the `Agent` and provide handles to files to the API tier on `GET` requests that the Genie application completely owns (or should own) this directory and its entire tree. As such the structure should never match these conditions and they're exceptional (e.g. someone logs into the server and creates a file on disk)

One weird edge case could be that currently this service only handles writing but if a job at runtime does file renames/deletions/etc things could get dicey. We probably need to think more about this case. Not sure if the watcher implementation on the Agent is set up to catch these events from the FS or not.

tgianos commented on genie-common-internal/src/main/java/com/netflix/genie/common/internal/dto/v4/logs/JobLogFile.java:
```diff
@@ -0,0 +1,65 @@
+/*
```
Perhaps semantic but seems reasonable enough. If we do this the whole service should probably be named `JobFileService` instead of `LogService` for consistency I would think

tgianos commented on genie-web/src/main/java/com/netflix/genie/web/services/impl/DiskLogServiceImpl.java:
```diff
@@ -0,0 +1,336 @@
+     * @param jobsDir The job directory resource
+     * @throws IOException When the job directory can't be created or isn't a directory
+     */
+    public DiskLogServiceImpl(final Resource jobsDir) throws IOException {
```
I'm not sure I understand the difference you're trying to establish between the two. Instead of `Log` think `output` or `working directory` if you'd like. In V3 these are written directly and served directly off local disk. If we put the logs that are uploaded from somewhere else in a different location than this existing job directory space it complicates the logic to return log files in response to API requests via the V3 API.

I could see an argument for a location where running job files are kept somewhere and finished jobs are placed somewhere else but i'm not sure what the value in the distinction is here.

tgianos commented on genie-web/src/main/java/com/netflix/genie/web/services/impl/DiskLogServiceImpl.java:
```diff
@@ -0,0 +1,336 @@
+        if (Files.notExists(dir)) {
+            Files.createDirectories(dir);
+        } else if (!Files.isDirectory(dir)) {
+            // TODO: Maybe a more specific exception here? Should be runtime maybe as there's no recovery?
```
I can't see anything using this service other than other pieces of the Genie codebase. I don't expect any users to interact or upload files directly.

tgianos commented:
@mprimi @sumitnetflix I renamed a bunch of files per earlier feedback

I added an initial implementation of the grpc JobFileSyncService. It's very rough but looking for feedback so I committed it. No tests yet.

sumitnetflix commented on genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobFileSyncServiceImpl.java:
```diff
@@ -0,0 +1,399 @@
+    private final JobFileService jobFileService;
+    private final TaskScheduler taskScheduler;
+
+    private final Map<String, GenieJobFileSyncObserver> jobSyncRequestObservers = Maps.newConcurrentMap();
```
Instead of having  `Maps.newConcurrentMap();`, would it be better to inject this as a dependency. It might help in testing. I had a similar map of observers for `JobKillService` and I was thinking of creating a interface and injecting its Impl

tgianos replied:
I don't agree with this. This is an internal construct for keeping state important to this particular implementation. If you provide a map externally you're effectively risking manipulation of that Map by outside sources, not to mention exposing internal implementation and breaking encapsulation. No one else should really care that you're keeping this map it's just for internal housekeeping for this class. If you really need access for it for testing you can provide a `protected` or `visablefortesting` getter to check the contents of the map.

mprimi commented on genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobFileSyncServiceImpl.java:
```diff
@@ -0,0 +1,609 @@
+
+        private void handleBeginSync(final BeginSync beginSync) throws IOException {
+            // Handle the begin message provided we're still waiting for it
+            if (this.waitingForBeginMessage) {
```
Could add a WARN if this condition is not true.

mprimi commented on genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobFileSyncServiceImpl.java:
```diff
@@ -0,0 +1,609 @@
+            final String messageId = deleteFile.getId();
+            log.debug("Received file delete message {} in job {}", messageId, this.jobId);
+
+            if (this.waitingForBeginMessage) {
```
This block appears in 3 methods, could be pulled out for less duplication and to make these methods simpler.

tgianos replied:
I left it in each due to wanting unique log messages that had specific details. Can change it later if we find it a burden. The logic to check whether to send the ack is already shared in one spot they all call.

mprimi commented on genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobFileSyncServiceImpl.java:
```diff
@@ -0,0 +1,609 @@
+            }
+        }
+    }
+}
```
This protocol is stateful on the server, if it gets any more complex we may consider using spring statemachine to encode it.

tgianos replied:
I agree we may want to do that or perhaps should. Outside the scope of this PR but it sounds like it might be the right thing to do.

tgianos commented on genie-web/src/main/java/com/netflix/genie/web/services/JobFileService.java:
```diff
@@ -0,0 +1,106 @@
+        final byte[] data
+    ) throws IOException;
+
+//    Optional<String> readLogFile(
```
Yeah it can be removed. I left it in cause I spent a bunch of time on it but eventually decided to punt for now and return a resource reference and allow callers to do with it as they need. Left the code in in case I wanted to use it again. I'll save it as a gist or something somewhere for my own use if I want to revisit.

tgianos commented on genie-proto/src/main/proto/genie.proto:
```diff
@@ -213,3 +213,83 @@ message ChangeJobStatusResponse {
+message ResetSync {
+}
+
+message UploadAcknowledgement {
```
Yeah I like this it is cleaner than what I have. I'll make the change.

tgianos commented on genie-web/src/main/java/com/netflix/genie/web/properties/JobFileSyncRpcProperties.java:
```diff
@@ -0,0 +1,37 @@
+@Getter
+@Setter
+public class JobFileSyncRpcProperties {
+    // TODO: Document properties
```
Yes I should. I forgot. Will add.

tgianos commented on genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobFileSyncServiceImpl.java:
```diff
@@ -0,0 +1,609 @@
+        //       for the same job. Just putting this here as temporary logic till we figure that out
+        final JobFileSyncObserver observer = this.jobSyncRequestObservers.get(jobId);
+        if (observer != null) {
+            // Cleanup will invoke the cleanup callback and remove it from the map
```
Going to hold off on modifying this until we finalize how we want to identify and remove things from map

tgianos commented:
@mprimi I've made changes based on your feedback and added one more commit at the end for a misspelling in the proto file. If you want to take one last look let me know otherwise I'll merge this tonight/tomorrow after builds

sumitnetflix commented on genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobFileSyncServiceImpl.java:
```diff
@@ -0,0 +1,612 @@
+     * received are ignored
+     * - Only one {@link ResetSync} message is sent no matter how many other messages that aren't {@link BeginSync}
+     * are received before the client sends a {@link BeginSync}
+     * 3. Once the {@link BeginSync} message is received processing continues as per the healthy workflow
```
👍 for the comment. Very helpful
